### PR TITLE
Add bias correction in Adam & Lamb for C++ frontend & python frontend

### PR DESCRIPTION
**Pull request by wschin**
Adam with bias correction is used many places ([an example](https://github.com/NVIDIA/apex/blob/80b90b9d43e343c7d4df7f6081a4ceaee2d3c348/csrc/multi_tensor_adam.cu#L98)), so we implement it in ORT.

In the first commit, we change Adam to always use bias correction. In the next commit, we will allow user to enable it or not.

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
Include @Tixxx in the conversation here.... 

How many variation of Adam do we see in the market? 
Adam vs. AdamW
With bias correction vs. without  bias correct
Apply Weight decay before vs. after weight update

For the latter 2 variations, we should make them attributes...

**Comment by SherlockNoMad** [comment]
Please include training tab if the future PRs.

**Comment by Tixxx** [comment]
> Include @Tixxx in the conversation here....
> 
> How many variation of Adam do we see in the market?
> Adam vs. AdamW
> With bias correction vs. without bias correct
> Apply Weight decay before vs. after weight update
> 
> For the latter 2 variations, we should make them attributes...

Besides Adam and Adamw, I have seen Adamax where the v term is replaced by an infinity norm, i think it was used in training mt-dnn at some point.
For the bias-correction and different weight decays logics, I agree they should be attributes for the optimizer.

File: orttraining/orttraining/models/bert/main.cc [hunk 1]
```diff
@@ -312,10 +317,13 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
     float beta = flags["beta"].as<float>();
     float lambda = flags["lambda"].as<float>();
     float epsilon = flags["epsilon"].as<float>();
+    int64_t do_bias_correction = flags["do_bias_correction"].as<int64_t>();
     float ratio_min = flags["ratio_min"].as<float>();
     float ratio_max = flags["ratio_max"].as<float>();
     ORT_RETURN_IF_NOT(alpha >= 0.f && alpha <= 1.f, "alpha is not in valid range [0.0, 1.0]");
     ORT_RETURN_IF_NOT(beta >= 0.f && beta <= 1.f, "alpha is not in valid range [0.0, 1.0]");
+    ORT_RETURN_IF_NOT(do_bias_correction == 0 || do_bias_correction == 1, "Bias correction can be either 0 or 1.");
+    ORT_RETURN_IF_NOT(epsilon >= 0.f, "epsilon should be non-negative.");
```
Comment (Tixxx) [inline orttraining/orttraining/models/bert/main.cc]: nit: same as the line below

**Comment by Tixxx** [comment]
also expose this to python api?

**Comment by wschin** [comment]
> also expose this to python api?

Yeah. I am working on it.

File: orttraining/orttraining/core/graph/optimizer/lamb_optimizer_builder.cc [hunk 2]
```diff
@@ -93,6 +93,7 @@ Status LambOptimizerBuilder::Build(
   std::vector<float> epsilon;
   float ratio_min = -std::numeric_limits<float>::infinity();
   float ratio_max = std::numeric_limits<float>::infinity();
+  int64_t do_bias_correction = 0;
```
Comment (Tixxx) [inline orttraining/orttraining/core/graph/optimizer/lamb_optimizer_builder.cc]: should the changes here be in adam builder?

File: orttraining/orttraining/core/graph/optimizer_builder.h [hunk 3]
```diff
@@ -43,7 +43,10 @@ class OptimizerBuilder {
  public:
   OptimizerBuilder(const std::string& name, const std::vector<std::string>& attribute_names = {})
       : name_(name),
-        attr_names_(attribute_names) {}
+        attr_names_(attribute_names) {
+
+
+        }
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/optimizer_builder.h]: revert?

File: orttraining/orttraining/core/graph/gradient_schema_defs.cc [hunk 4]
```diff
@@ -473,6 +478,11 @@ void RegisterGradientSchemas() {
           "Small scalar to avoid dividing by zero.",
           AttributeProto::FLOAT,
           1e-8f)
+      .Attr(
+          "do_bias_correction",
+          "Compute unbiased 1st and 2nd momentums.",
+          AttributeProto::INT,
+          static_cast<int64_t>(1))
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_schema_defs.cc]: I am wondering if we need to make this a enum?
Just like apex does... 

typedef enum{
  ADAM_MODE_0   =0, // L2 regularization mode
  ADAM_MODE_1   =1  // Decoupled weight decay mode(AdamW)
} adamMode_t;

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_schema_defs.cc]: Is TJ's weight decay variation going to be another flag ?

---
In reply to: [397589273](https://github.com/microsoft/onnxruntime/pull/3301#discussion_r397589273) [](ancestors = 397589273)

File: orttraining/orttraining/python/ort_trainer.py [hunk 5]
```diff
@@ -420,14 +420,22 @@ def create_ort_training_session_with_optimizer(model, device, training_optimizer
...
         weights_to_train.add(initializer.name)
         if map_optimizer_attributes is not None:
             attributes = map_optimizer_attributes(initializer.name)
-            optimizer_attributes_map[initializer.name] = attributes
+            optimizer_attributes_map[initializer.name] = {}
+            optimizer_int_attributes_map[initializer.name] = {}
+            for k, v in attributes.items():
+                if isinstance(v, float):
+                    optimizer_attributes_map[initializer.name][k] = v
+                else:
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/ort_trainer.py]: >else [](start = 16, length = 4)

make explicit else if int here

**Comment by SherlockNoMad** [comment]
> TEST(OptimizerTest, AdamOptimizerTest) {

Consider move optimizer test out of this file 

gradient_op_tests.cc

---
Refers to: orttraining/orttraining/test/gradient/gradient_ops_test.cc:1182 in 18f3cf2. [](commit_id = 18f3cf257bcdb59f1df56431e547c71c6fa88058, deletion_comment = False)

File: orttraining/orttraining/models/bert/main.cc [hunk 6]
```diff
@@ -136,6 +136,11 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
       ("beta", "Adam/Lamb beta parameter", cxxopts::value<float>()->default_value("0.999"))
       ("lambda", "Adam/Lamb lambda parameter", cxxopts::value<float>()->default_value("0.01"))
       ("epsilon", "Adam/Lamb epsilon parameter", cxxopts::value<float>()->default_value("1e-6"))
+      ("do_bias_correction",
+        "A flag controls if Adam/Lamb should do bias correction. "
+        "Default is false, which means no bias correction. "
+        "Use true to enable bias correction.",
+        cxxopts::value<bool>()->default_value("false"))
```
Comment (SherlockNoMad) [inline orttraining/orttraining/models/bert/main.cc]: >false [](start = 47, length = 5)

the schema is default 1 

shall they match ?

File: orttraining/orttraining/training_ops/cuda/optimizer/adam.h [hunk 7]
```diff
@@ -38,6 +39,7 @@ class AdamOptimizer final : public CudaKernel {
     info.GetAttrOrDefault("beta", &beta_, 0.999f);
     info.GetAttrOrDefault("lambda", &lambda_, 0.0f);
     info.GetAttrOrDefault("epsilon", &epsilon_, 1e-8f);
+    ORT_ENFORCE(info.GetAttr<int64_t>("do_bias_correction", &do_bias_correction_).IsOK(), "Missing/Invalid do_bias_correction");
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/adam.h]: >do_bias_correction_ [](start = 61, length = 19)

constructor is a good place to convert to bool

File: orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h [hunk 8]
```diff
@@ -34,6 +41,7 @@ class AdamOptimizer final : public OpKernel {
   float beta_;
   float lambda_;
   float epsilon_;
+  int64_t do_bias_correction_;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h]: >int64_t [](start = 2, length = 7)

bool ?

**Comment by SherlockNoMad** [comment]
>       }

step is not updated in the case of overflow.

is this correct?

---
Refers to: orttraining/orttraining/training_ops/cuda/optimizer/adam.cc:110 in 18f3cf2. [](commit_id = 18f3cf257bcdb59f1df56431e547c71c6fa88058, deletion_comment = False)

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.h [hunk 9]
```diff
@@ -31,6 +32,7 @@ class LambOptimizer final : public CudaKernel {
   std::vector<float> epsilon_;
   float ratio_min_;
   float ratio_max_;
+  int64_t do_bias_correction_;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.h]: >int64_t [](start = 2, length = 7)

when it come to kernel private member, I feel it's better to store as bool

File: orttraining/orttraining/core/graph/gradient_schema_defs.cc (hunk 4 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_schema_defs.cc]: Nevermind, this flag should stand alone ... 

---
In reply to: [397590509](https://github.com/microsoft/onnxruntime/pull/3301#discussion_r397590509) [](ancestors = 397590509,397589273)

File: orttraining/orttraining/core/graph/optimizer_builder.h (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/core/graph/optimizer_builder.h]: Sure.

File: orttraining/orttraining/core/graph/gradient_schema_defs.cc (hunk 4 above)
Comment (wschin) [reply orttraining/orttraining/core/graph/gradient_schema_defs.cc]: Yes. Weight decay is another flag. They are independent and compatible with each other.

File: orttraining/orttraining/models/bert/main.cc (hunk 6 above)
Comment (wschin) [reply orttraining/orttraining/models/bert/main.cc]: I feel it's better not to change the behavior of C++ frontend. It'd be surprise if the same command produces very different results.

File: orttraining/orttraining/training_ops/cuda/optimizer/adam.h (hunk 7 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/optimizer/adam.h]: Sure.

File: orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h (hunk 8 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cpu/optimizer/optimizers.h]: Ok.

**Comment by wschin** [comment]
> > ```
> >   }
> > ```
> 
> step is not updated in the case of overflow.
> 
> is this same as pytorch/nvbert scirpt?
> 
> Refers to: orttraining/orttraining/training_ops/cuda/optimizer/adam.cc:110 in [18f3cf2](https://github.com/microsoft/onnxruntime/commit/18f3cf257bcdb59f1df56431e547c71c6fa88058). [](commit_id = [18f3cf2](https://github.com/microsoft/onnxruntime/commit/18f3cf257bcdb59f1df56431e547c71c6fa88058), deletion_comment = False)

Yeah, global_step is increased only if gradient is finite as shown at https://github.com/NVIDIA/DeepLearningExamples/blob/e663119b51c9460c1632d2169b00fc014c29aea1/PyTorch/LanguageModeling/BERT/run_pretraining.py#L402

File: orttraining/orttraining/python/ort_trainer.py (hunk 5 above)
Comment (wschin) [reply orttraining/orttraining/python/ort_trainer.py]: Ok.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.h (hunk 9 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/optimizer/lamb.h]: Sure.

File: orttraining/orttraining/training_ops/cuda/optimizer/lamb.h [hunk 10]
```diff
@@ -20,6 +20,11 @@ class LambOptimizer final : public CudaKernel {
     epsilon_ = info.GetAttrsOrDefault("epsilon", std::vector<float>(1024, 1e-6f));
     ORT_ENFORCE(info.GetAttr<float>("ratio_min", &ratio_min_).IsOK(), "Missing/Invalid 'ratio_min' attribute value");
     ORT_ENFORCE(info.GetAttr<float>("ratio_max", &ratio_max_).IsOK(), "Missing/Invalid 'ratio_max' attribute value");
+
+    int64_t tmp_flag = static_cast<int64_t>(0);
+    ORT_ENFORCE(info.GetAttr<int64_t>("do_bias_correction", &tmp_flag).IsOK(), "Missing/Invalid do_bias_correction");
+    ORT_ENFORCE(tmp_flag == 0 || tmp_flag == 1, "do_bias_correction must be either 0 or 1.");
+    do_bias_correction_ = tmp_flag != 0 ? true : false;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/lamb.h]: nit. tmp_flag != 0 is already a bool
also in other construcutor...
